File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/TemporalEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.temporal.Temporal;
import java.util.Iterator;
import java.util.List;

public class TemporalEntityTransform extends EntityTransform<Temporal> {

    private static class TemporalEntityTransformHolder {
        private static final TemporalEntityTransform INSTANCE = new TemporalEntityTransform();
    }
    // global access point
    public static TemporalEntityTransform getInstance() {
        return TemporalEntityTransform.TemporalEntityTransformHolder.INSTANCE;
    }
    private TemporalEntityTransform() {
        super(EntityType.DATETIME, Temporal.class);
    }


    @Override
    public void serializeEntity(Temporal entity, List<Byte> bytes) {
        bytes.add(EntityType.DATETIME.getCode());
        ZonedDateTime dt;
            // A ZonedDateTime carries a real offset, so it must be *converted* to UTC
            // (withZoneSameInstant), not relabelled (withZoneSameLocal) - the latter keeps the
            // wall-clock reading and silently moves the instant by that offset.
            // A LocalDateTime carries no offset and is already a UTC instant by contract (see
            // TimeProvider), so stamping the zone on is correct there.
            if (entity instanceof ZonedDateTime zdt)
                dt = zdt.withZoneSameInstant(ZoneOffset.UTC);
            else{
                dt = ((LocalDateTime)entity).atZone(
                        ZoneOffset.UTC
                );
            }
        var millis = dt.toInstant().toEpochMilli();
        var nanos = dt.getNano();
        writeLong(millis, bytes);
        writeInt(nanos, bytes);
    }

    @Override public Temporal  deserializeEntity(Iterator<Byte> bytes) {
        var millis = readLong(bytes);
        var nanos = readInt(bytes);

        return LocalDateTime.ofEpochSecond(Double.valueOf(millis / 1000.).longValue(), nanos, ZoneOffset.UTC);
    }
}

File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/TemporalEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType
import com.loudsight.useful.helper.ClassHelper
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.temporal.Temporal

object TemporalEntityTransform : EntityTransform<Temporal>(EntityType.DATETIME, Temporal::class) {

    override fun serializeEntity(entity: Temporal, bytes: MutableList<Byte>) {
        bytes.add(EntityType.DATETIME.code)
        val zdt: ZonedDateTime =
            if (entity is ZonedDateTime)
                entity.withZoneSameLocal(ZoneOffset.UTC)
            else (entity as LocalDateTime).atZone(
                ZoneOffset.UTC
            )
        val millis = zdt.toInstant().toEpochMilli()
        val nanos = zdt.nano
        writeLong(millis, bytes)
        writeInt(nanos, bytes)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Temporal {
        val millis = readLong(bytes)
        val nanos = readInt(bytes)

        return LocalDateTime.ofEpochSecond(millis / 1000, nanos, ZoneOffset.UTC)
    }


    override fun canTransform(entity: Any): Boolean {
        return entity is Temporal
    }

}